pie/models/decoder.py

```diff
@@ -328,35 +328,100 @@ def predict_max(self, enc_outs, lengths,
-            emb = emb.unsqueeze(0)
+                emb = torch.cat([emb, context], dim=1)  # Tensor(batch_size x (emb_size+context_size))
+
+            # Run rnn
+            emb = emb.unsqueeze(0)  # Tensor(1 x batch_size x emb size(+context))
+
+            # hidden is gonna be reused by the next iteration
+            #   outs is specific to the current run
             outs, hidden = self.rnn(emb, hidden)
+            # Hidden : Tensor(1 x batch_size x emb_size)
+
             outs, _ = self.attn(outs, enc_outs, lengths)
             outs = self.proj(outs).squeeze(0)
-            # get logits
+
+            # Get logits
             probs = F.log_softmax(outs, dim=1)
-            # sample and accumulate
-            score, inp = probs.max(1)
-            hyps.append(inp.tolist())
-            mask = mask * (inp != eos).long()
-            score = score.cpu()
-            score[mask == 0] = 0
-            scores += score
+
+            # Sample and accumulate
+            #  Score are the probabilities
+            #  Inp are the new characters (as int) we are adding to our predictions
+            score, inp = probs.max(1)  # (Tensor(batch_size, dtype=float), Tensor(batch_size, dtype=int))
+
+            # We create a mask of value that are not ending the string
+            non_eos = (inp != eos)  # Tensor(batch_size, dtype=bool)
+
+            # Using this mask, we retrieve the Indexes of items that are not EOS
+            #  nonzero() returns a tuple where the first item is a Tensor
+            #  with the indexes. It can be use as a selector for other tensors (see below)
+            keep, *_ = torch.nonzero(non_eos, as_tuple=True)  # Tensor(dtype=int)
+
+            # Add new chars to hypotheses
+            #   We prepare a list the size of the output, filling it with EOS
+            #   We then iterate over score and inp (same sized tensors) and add to our
+            prediction_run_output = [eos for _ in range(batch)]
+
+            for ind, (hyp, sc) in enumerate(zip(inp.tolist(), score.tolist())):
+                # To add the hypothesis to the right word, we use the associated Tensor->Batch indexes
+                prediction_run_output[tensor_to_batch_indexes[ind]] = hyp
+                # Probability are only added if they are not EOS chars
+                if hyp != eos:
+                    scores[tensor_to_batch_indexes[ind]] += sc
+
+            # We add this new output to the final hypothesis
+            hyps.append(prediction_run_output)
+
+            # If there nothing else than EOS, it's the end of the prediction time
+            if True not in non_eos:
+                break
+
+            # Otherwise, we update the tensor_to_batch_indexes by transferring
+            #   the current associated index with the new indexes
+            tensor_to_batch_indexes = {
+                elem: tensor_to_batch_indexes[former_index]
+                for elem, former_index in enumerate(keep.tolist())
+            }
```

Comment:
See above. The keys are always  0 to n (remaining items) so the dictionary is redundant. I'd rather go for a tensor that lets you index into the output tensors.

Reply:
Done.